src/main/scala/calc/Compiler.scala

```diff
@@ -62,21 +61,101 @@ object Compiler {
   }
 
   /** Compile an expression tree into an IR `Tree`, which is an expression
-   *  that evaluates to the result of the tree.
-   *
-   *  This is the main method you have to implement.
-   */
+    *  that evaluates to the result of the tree.
+    *
+    *  This is the main method you have to implement.
+    */
   def compileExpr(tree: Tree): irt.Tree = {
-    // TODO
     implicit val pos = tree.pos
 
-    tree match {
-      case Literal(value) =>
-        irt.DoubleLiteral(value)
+    val foreignEnv = Foreign.staticJavaMethods(
+      clsName = "java.lang.Math",
+      importedMethods = Map(
+        "sin__D__D" -> ("sin" -> TFun(1)),
+        "cos__D__D" -> ("cos" -> TFun(1)),
+        "sqrt__D__D" -> ("sqrt" -> TFun(1)),
+        "pow__D__D__D" -> ("pow" -> TFun(2))
+      )
+    )
 
-      case _ =>
-        throw new Exception(
-            s"Cannot yet compile a tree of class ${tree.getClass}")
+    implicit val env = Typer.emptyEnv ++ foreignEnv
+    expr(Typer.inferType(tree))
+  }
+
+  def expr(t: TreeT): irt.Tree = {
+    t match {
+      case t: IdentT => ident(t)
+      case t: LiteralT => literal(t)
+      case t: BinaryOpT => binaryOp(t)
+      case t: LetT => letBinding(t)
+      case t: IfT => ifElse(t)
+      case t: ClosureT => closure(t)
+      case t: CallT => call(t)
+      case t: ForeignCallT => foreignCall(t)
+    }
+  }
+
+  def ident(t: IdentT): irt.Tree = { implicit val pos = t.pos
+    irt.VarRef(irt.Ident(t.name))(t.tpe.irtype)
+  }
+
+  def literal(t: LiteralT) = { implicit val pos = t.pos
+    irt.DoubleLiteral(t.value)
+  }
+
+  def binaryOp(t: BinaryOpT) = { implicit val pos = t.pos
+    val irOp = operatorToIR(t.op) getOrElse (throw UnknownOperator(pos, t.op))
+    irt.BinaryOp(irOp, expr(t.lhs), expr(t.rhs))
+  }
+
+  def letBinding(t: LetT) = { implicit val pos = t.pos
+    val bindingRhs = expr(t.value)
+    val binding = irt.VarDef(irt.Ident(t.name.name), t.value.tpe.irtype, false, bindingRhs)
+    val body = expr(t.body)
+    irt.Block(List(binding, body))
+  }
+
+  def ifElse(t: IfT) = { implicit val pos = t.pos
+    val cond = irt.BinaryOp(irt.BinaryOp.!==, expr(t.cond), irt.DoubleLiteral(0))
+    val thenp = expr(t.thenp)
+    val elsep = expr(t.elsep)
+    irt.If(cond, thenp, elsep)(t.tpe.irtype)
+  }
+
+  def closure(t: ClosureT) = { implicit val pos = t.pos
+    val paramBoxed = t.params map { ident =>
+      irt.ParamDef(irt.Ident(ident.name), irtpe.AnyType, false, false)
+    }
+    val paramUnbox = t.params map { ident =>
+      val varRef = irt.VarRef(irt.Ident(ident.name))(irtpe.AnyType)
+      val varUnbox = irt.Unbox(varRef, 'D')
+      irt.VarDef(irt.Ident(ident.name), irtpe.DoubleType, false, varUnbox)
```

Comment:
Uh oh. You're using the same `ident.name` for the `ParamDef` and the local `VarDef`. That's bad, as they live in the same scope. I'm surprised it did not blow up already. You should use a different (generated) name for the `ParamDef` and its `VarRef`.